assets/src/data/persistence/extrinsic.ts

```diff
@@ -52,10 +52,35 @@ export const readGlobalUserState = async (
     if (lastSet()) {
       await lastSet();
     }
-    const serverUserState = await readGlobal(keys);
-    // merge server state with result
-    if ((serverUserState as any).type !== 'ServerError') {
-      result = serverUserState;
+    let cacheMissingRequestedKey = false;
+    if (keys) {
+      const lastchacheTimeStamp = userStateCache['timestamp'];
+      //If cache is not older than 20 sec then lets fetch the data from cache
+      if (Date.now() - lastchacheTimeStamp < 20000) {
+        keys.forEach((key) => {
+          if (!userStateCache[key]) {
+            //if cache does not have any of the requested keys, we should make the server call
+            cacheMissingRequestedKey = true;
+          }
+          result[key] = userStateCache[key];
+        });
+      } else {
+        const serverUserState = await readGlobal(keys);
+        // merge server state with result
+        if ((serverUserState as any).type !== 'ServerError') {
+          result = serverUserState;
+        }
+      }
+    } else {
+      result = userStateCache;
+    }
+    if (cacheMissingRequestedKey) {
+      //if cache does not have any of the requested keys, we should make the server call
+      const serverUserState = await readGlobal(keys);
```

Comment:
since this is repeated, probably better to just manage a single flag; 
maybe rename `cacheMissingRequestedKey` to `refreshFromServer` and then if it's because of the time or because it was missing a key, just do it once at the bottom (here) `if (refreshFromServer) `